### PR TITLE
Exposing headers to context object and inside postgreSQL as req.headers

### PR DESCRIPTION
**Pull request by mikeburgh**
A solution for #299 which adds the req.headers using set_config.

Only thought I had was do we want to hide these behind an option, so that by default headers are not pushed into postgres ?

### CODE REVIEW

**Comment by benjie** [comment]
Thanks for the PR 🙏

There needs to be a whitelist, really - or at least some sanity checks. Currently someone could send through an unlimited number of arbitrarily large header keys with arbitrary large values. As you say, there should also be an option to turn this on/off. I'm not sure what performance/memory concerns set_config has.

At least req.headers already does some normalisation for us 🙂

**Comment by benjie** [comment]
I also don't like the idea of Options getting too many more keys; can we add a property that would include headers and other similar properties, e.g. "context". Or maybe just pass through the whole req object?

**Comment by mikeburgh** [comment]
With the option, we could do a single context option, and give it a json struct for what you want on the context (there maybe a better format)

--context {req:{headers:true}} // would apply all the headers to the context and set_config

--context {req:{headers:'host,user-agent'}} // would apply the host and user-agent only headers to the context and set_config

--context {req:{body:true,headers:true}} // would apply all headers and body to the context and set_config

--context {req:true, custom:{abc:123,xyz:456}} // would apply the whole req object, and the custom options to the context and set_config

In this way, you could expand it if there were other things that came up that people wanted set, and you also get custom settings in the command line (I know you can apply custom ones via pgSettings, but I don't think you can get to that from the cli).

If we go this path, then we don't really need to do the whitelist approach as people will be in charge of that via the option (assuming they don't just do req:true).. and perhaps we limit each set_config item to 1000chars just to be safe ?

**Comment by benjie** [comment]
Hi @mikeburgh, I've thought about this some more and I think we want to approach this from a different, less specific, angle.

We have a library option called 'pgSettings' already for passing through static settings. I'm thinking we should change this into a more flexible API by coercing it into an array (i.e. wrapping it with an array if it isn't already an array), where each element is either:

- an object: gets merged into the list of settings to set
- string that starts with a `{`: JSON.parse it (so we can accept it on the command line) and treat it as an object above
- function: gets called with a few interesting thing (e.g. `context`, which should have the `req` object on it), returns an object to be merged into the settings to set
- simple string: is a reference to a list of built in functions:
  - `req.headers` - `({context: {req}}) => mapKeys(req.headers, (value, key) => 'req.headers.' + key)` (but with some protections for number of keys/length of values)
  - `req.user` - `({context: {req}}) => mapKeys(req.user, (value, key) => 'req.user.' + key)` (with similar protections)
  - etc

then we can also expose this on the command line. e.g.

````
postgraphql --pgSettings req.headers --pgSettings req.user --pgSettings '{"a.b": "foo", "c.d": "bar"}'
```

Do you have any thoughts on this?

**Comment by chadfurman** [comment]
@benjie I like your idea.  I think this goes nicely with https://github.com/postgraphql/postgraphql/issues/501   I think

**Comment by benjie** [comment]
When using PostGraphile as a library this functionality is already available via `pgSettings` (though we only support the function and object forms). I'm going to leave this issue open as a reminder to add a CLI equivalent that does the above.

https://www.graphile.org/postgraphile/usage-library/#pgsettings-function

Anything that implements `req.headers` or similar will require significant scrutiny to avoid making the system vulnerable to attacks that set many or large headers.

File: src/postgraphql/withPostGraphQLContext.ts [hunk 1]
```diff
@@ -167,6 +172,11 @@ async function setupPgClientTransaction ({
     localSettings.set(`jwt.claims.${key}`, jwtClaims[key])
   }
 
+  // If we have headers!
+  for (const key of Object.keys(reqHeaders)) {
+    localSettings.set(`req.headers.${key}`, reqHeaders[key])
```
Comment (chadfurman) [inline src/postgraphql/withPostGraphQLContext.ts]: Is this safe?  do we use localSettings in a way where we do not expect escaping?  Are these all type-cast to text?

**Comment by benjie** [comment]
Replacing this with an issue: https://github.com/graphile/postgraphile/issues/769

**Pull request closed without merging**